### PR TITLE
[FLINK-33708][FLINK-33709] Introduce TraceReporter and use it to create checkpointing traces

### PR DESCRIPTION
**Pull request by pnowojski**
## What is the purpose of the change

This PR adds a concept of `Span` and `TraceReporter` and uses that to report checkpointing traces.

## Brief change log

Please check individual commit messages.

## Verifying this change

This change added tests and can be verified as follows:
- `CheckpointStatsTrackerTest#testSpanCreation` checks that a span with proper Checkpoint ID is created.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

File: docs/content/docs/ops/traces.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+## Reporter
+
+For information on how to set up Flink's trace reporters please take a look at the [trace reporters documentation]({{< ref "docs/deployment/trace_reporters" >}}).
+
+## System traces
+
+Flink reports traces listed below.
+
+The tables below generally feature 5 columns:
```
Comment (rkhachatryan) [inline docs/content/docs/ops/traces.md:75]: typo: feature**s** ?

File: docs/content/docs/ops/traces.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    <tr>
+      <td>endTs</td>
+      <td>Timestamp when the checkpoint has finished.</td>
+    </tr>
+    <tr>
+      <td>checkpointId</td>
+      <td>Id of the checkpoint.</td>
+    </tr>
+    <tr>
+      <td>checkpointSize</td>
```
Comment (rkhachatryan) [inline docs/content/docs/ops/traces.md:112]: Isn't it checkpoint**ed**Size?

File: docs/content/docs/ops/traces.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+---
+title: "Metrics"
```
Comment (rkhachatryan) [inline docs/content/docs/ops/traces.md:2]: Metrics -> Traces

(I see two identical pages currently in navigation)

File: docs/content/docs/ops/traces.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+* The "Scope" column describes what is that trace reported scope.
+
+* The "Name" column describes the name of the reported trace.
+
+* The "Attributes" column lists the names of all attributes that are reported with the given trace.
+
+* The "Description" column provides information as to what a given attribute is reporting.
+
+### Checkpointing and recovery
+
```
Comment (rkhachatryan) [inline docs/content/docs/ops/traces.md:86]: Should we clarify that the trace is reported only once the checkpoint is completed/failed?

File: docs/layouts/shortcodes/generated/akka_configuration.html [hunk 5]
```diff
@@ -98,6 +80,24 @@
...
+            <td style="word-wrap: break-word;">16</td>
+            <td>Integer</td>
+            <td>Max number of threads to cap factor-based parallelism number to.</td>
+        </tr>
+        <tr>
+            <td><h5>pekko.remote-fork-join-executor.parallelism-min</h5></td>
+            <td style="word-wrap: break-word;">8</td>
+            <td>Integer</td>
+            <td>Min number of threads to cap factor-based parallelism number to.</td>
+        </tr>
```
Comment (rkhachatryan) [inline docs/layouts/shortcodes/generated/akka_configuration.html:100]: Unnecessary change?

File: docs/layouts/shortcodes/generated/trace_configuration.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<table class="configuration table table-bordered">
```
Comment (rkhachatryan) [inline docs/layouts/shortcodes/generated/trace_configuration.html:1]: I couldn't find this section.
Shouldn't it be added explicitly to `config.md`?

**Comment by rkhachatryan** [comment]
One more thing, all the documentation changes only affect the English version.
Should we also update the Chinese version (with the same content in English)?

File: docs/content/docs/ops/traces.md (hunk 1 above)
Comment (pnowojski) [reply docs/content/docs/ops/traces.md:75]: 🤔 The sentence refers to `tables` plural, so `feature` is correct. But atm there is only a single table, so I'm not sure which way to go. I would lean towards keeping the plural version as it's future proof. WDYT?

File: docs/content/docs/deployment/config.md [hunk 7]
```diff
@@ -289,6 +289,15 @@ Enabling RocksDB's native metrics may cause degraded performance and should be s
 ----
 ----
 
+# Traces
```
Comment (rkhachatryan) [inline docs/content/docs/deployment/config.md:292]: NIT: This doesn't show up in TOC - it should be 2 or more `#` with the current settings.

As well as other, existing section (e.g. `History Server` is missing).

Settings can be fixed by adding to `docs/config.toml`:
```
[markup.tableOfContents]
startLevel = 1
```
Not sure if this change would be reasonable and whether it can be added as a hotfix to this PR.

File: docs/content/docs/ops/traces.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+* The "Name" column describes the name of the reported trace.
+
+* The "Attributes" column lists the names of all attributes that are reported with the given trace.
+
+* The "Description" column provides information as to what a given attribute is reporting.
+
+### Checkpointing
+
+Flink is reporting a single span trace for the whole checkpoint once checkpoint reaches a terminal state: COMPLETED or FAILED.
```
Comment (rkhachatryan) [inline docs/content/docs/ops/traces.md:87]: NIT: "Flink report a single span" ?

File: docs/content/docs/deployment/config.md (hunk 7 above)
Comment (pnowojski) [reply docs/content/docs/deployment/config.md:292]: Huh, good catch. I was curious a couple of times why the TOC seem broken, but never investigated it. I think this is a good fix, but let's ask @zentol  maybe.

Comment (zentol) [reply docs/content/docs/deployment/config.md:292]: I think this is intentional such that the main heading isn't part of the ToC.

> we shouldn't have more than a single Header lvl 1

correct, this is probably also why the Configuration ToC is so crap.

Comment (pnowojski) [reply docs/content/docs/deployment/config.md:292]: I've clicked through quite a lot of pages, and I didn't find anything too concerning or broken. One can argue that the extra level of indentation on most pages slightly reduces readability, as ToC is smaller or takes more space, but that's in an exchange of showing the title in the ToC. I think that's a fair trade.

And in those pages that have multiple headers lvl 1 ToC obviously is much much better.

**Comment by pnowojski** [comment]
Thanks for the review, merging.

Test failure was an unrelated issue on the master branch: FLINK-33480
